File: python/jaiku/jaiku.py
#!/usr/bin/env python

'''Interface with Jaiku API'''

__author__ = 'Andrew Pennebaker (andrew.pennebaker@gmail.com)'
__date__ = '28 Jun 2007'
__copyright__ = 'Copyright 2007 Andrew Pennebaker'
__version__ = '0.0.1'
__URL__ = 'http://snippets.dzone.com/posts/show/4223'

import configreader

import sys
import getopt
import urllib2
import xmlrpclib

STATUS_MODE = 'STATUS'
VIEW_MODE = 'VIEW'


def set_status(settings, status):
    '''Set Jaiku status'''

    s = xmlrpclib.ServerProxy(settings['xmlrpcurl'])

    calldata = {
        'user': settings['username'],
        'personal_key': settings['personalkey'],
        'message': status,
        'location': settings['location']
    }

    try:
        s.presence.send(calldata)
    except:
        raise Exception('Could not connect.')


def view_status(settings):
    '''View Jaiku status'''

    item = settings['itemdelimeter']
    t1 = settings['titledelimeter1']
    t2 = settings['titledelimeter2']

    try:
        instream = urllib2.urlopen(
            settings['feedurlstart'] +
            settings['username'] +
            settings['feedurlend']
        )

        for line in instream:
            if item in line:
                break

        title = instream.readline()

        instream.close()

        status = title[title.index(t1) + len(t1):title.index(t2)]

        return status

    except IOError:
        raise Exception('Could not connect.')


def usage():
    '''Print usage message'''

    print('Usage: %s [options]' % (sys.argv[0]))
    print('\nWithout any options, uses status mode.')
    print('Leftover args are concatenated to form message.')
    print('\n-u|--username <username> specified in jaiku.conf')
    print('-p|--personal-key <key>')
    print('-l|--location <location>')
    print('-s|--status mode')
    print('-v|--view status')
    print('-c|--config <configfile>')
    print('-h|--help')

    sys.exit()


def main():
    '''Jaiku CLI'''

    system_args = sys.argv[1:]

    mode = STATUS_MODE

    settings = {
        'config': 'jaiku.conf',
        'xmlrpcurl': 'http: //api.jaiku.com/xmlrpc',
        'feedurlstart': 'http: //',
        'feedurlend': '.jaiku.com/feed/atom',
        'itemdelimeter': '<entry>',
        'titledelimeter1': '<title>',
        'titledelimeter2': '</title>',
        'username': 'mcandre',
        'personalkey': '',
        'location': ''
    }

    optlist, args = [], []

    try:
        optlist, args = getopt.getopt(
            system_args,
            'u:p:l:svc:h',
            [
                'username=',
                'personal-key=',
                'location=',
                'status',
                'view',
                'config=',
                'help'
            ]
        )
    except getopt.GetoptError:
        usage()

    for option, value in optlist:
        if option == '-c' or option == '--config':
            settings['config'] = value

    try:
        configreader.load(open(settings['config'], 'r'), settings)
    except IOError:
        pass

    for option, value in optlist:
        if option == '-h' or option == '--help':
            usage()

        elif option == '-u' or option == '--username':
            settings['username'] = value
        elif option == '-p' or option == '--personal-key':
            settings['personalkey'] = value
        elif option == '-l' or option == '--location':
            settings['location'] = value
        elif option == '-s' or option == '--status':
            mode = STATUS_MODE
        elif option == '-v' or option == '--view':
            mode = VIEW_MODE

    if mode == STATUS_MODE:
        if len(args) < 1:
            usage()

        message = ' '.join(args)

        set_status(settings, message)
    elif mode == VIEW_MODE:
        print(view_status(settings))

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        pass

File: ruby/jaiku/jaiku.rb
#!/usr/bin/env ruby

# Author:: Andrew Pennebaker
# Copyright:: Copyright 2007-2008 Andrew Pennebaker
#
# == Synopsis
#
# jaiku.rb: updates and views Jaiku status messages
#
# == Usage
#
# jaiku [OPTIONS] <status>
#
# status must not be more than 140 characters
#
# --help, -h:
#    show help
#
# --debug, -d:
#    enable debugging
#
# --user, -u <user>
#
# --view, -v:
#    view mode
#
# --post, -p:
#    post mode (default)
#
# --list-commands, -l:
#    command mode

require "getoptlong"
require "rdoc/usage"

require "net/http"

require "rubygems"

# prefer faster json parser
begin
  require "json/ext"
rescue LoadError => e
  require "json"
end

require "highline/import"

$MAX_STATUS_LENGTH = 140

def self.load_settings(stream)
  settings = {}

  YAML::load(stream).each { |description, command|
    settings[description] = command
  }

  settings
end

def update(settings)
  debug = settings["debug"]
  domain = settings["post_domain"]
  api = settings["post_api"]
  status = settings["status"]
  username = settings["username"]
  password = settings["password"]
  key = settings["key"]

  raise "Status too long, shorten to #{$MAX_STATUS_LENGTH} characters" unless status.length <= $MAX_STATUS_LENGTH

  Net::HTTP.start(domain) { |http|
    request=Net::HTTP::Post.new(api)
    request.basic_auth(username, password)
    request.set_form_data({ "user" => username, "personal_key" => key, "method" => "presence.send", "message" => status })

    response = http.request(request)

    p response if debug

    message = response.message

    raise "Could not authenticate as #{username}" if message["Unauthorized"]
    raise "Could not connect" if message["Not Acceptable"]
  }
end

def view(settings)
  debug = settings["debug"]
  domain = settings["view_domain"]
  api = settings["view_api"]
  username = settings["username"]

  timeline = nil
  response = nil

  Net::HTTP.start(username + domain) { |http|
    response = http.get(api)

    p response if debug

    raise "Could not connect" if response.message["Not Acceptable"]

    timeline = response.body
  }

  stream = JSON.parse(timeline)["stream"]

  raise "No posts yet" if stream.length == 0

  status = stream[0]["title"]

  status
end

def main()
  mode = :post
  settings = {
    "debug" => false,
    "post_domain" => "api.jaiku.com",
    "post_api" => "/json",
    "view_domain" => ".jaiku.com",
    "view_api" => "/feed/json",
    "username" => "mcandre",
    "key" => 0
  }

  begin
    open("#{File.dirname($0)}/jaiku.yaml") { |file|
      settings = load_settings(file)
    }
  rescue Errno::ENOENT => e
    raise "Could not open settings file"
  end

  opts = GetoptLong.new(
    ["--help", "-h", GetoptLong::NO_ARGUMENT],
    ["--debug", "-d", GetoptLong::NO_ARGUMENT],
    ["--user", "-u", GetoptLong::REQUIRED_ARGUMENT],
    ["--view", "-v", GetoptLong::NO_ARGUMENT],
    ["--post", "-p", GetoptLong::NO_ARGUMENT],
    ["--list-commands", "-y", GetoptLong::NO_ARGUMENT]
  )

  begin
    opts.each { |option, value|
      case option
      when "--help"
        raise
      when "--debug"
        settings["debug"] = true
      when "--user"
        settings["username"] = value
      when "--view"
        mode = :view
      when "--post"
        mode = :post
      end
    }
  rescue
    RDoc::usage("Usage")
  end

  case mode
  when :view
    puts view(settings)
  when :post
    if ARGV.length < 1
      RDoc::usage("Usage")
    else
      settings["password"] = ask("Password: ") { |q| q.echo = false }
      settings["status"] = ARGV.join " "

      update(settings)
    end
  end
end

if __FILE__ == $0
  begin
    main
  rescue Timeout::Error, SocketError => e
    puts "Could not connect"
  rescue Interrupt => e
    nil
  end
end
